--- ./esc/src/lib/NssHttpClient/httpClientNss.cpp.fix7	2009-06-22 13:00:20.000000000 -0700
+++ ./esc/src/lib/NssHttpClient/httpClientNss.cpp	2009-06-22 13:05:07.000000000 -0700
@@ -64,8 +64,53 @@
 
 PSHttpResponse *HttpClientNss::httpSendChunked(char *host_port, char *uri, char *method, char *body,PSChunkedResponseCallback cb,void *uw,PRBool doSSL,int messageTimeout )
 {
-    
-    PSHttpServer server(host_port, PR_AF_INET);
+    char *pPort = NULL;
+    char *pPortActual = NULL;
+
+    char hostName[512];
+
+    /*
+     * Isolate the host name, account for IPV6 numeric addresses.
+     *
+     */
+
+    if(host_port)
+        strncpy(hostName,host_port,512);
+
+    pPort = hostName;
+    while(1)  {
+        pPort = strchr(pPort, ':');
+        if (pPort) {
+            pPortActual = pPort;
+            pPort++;
+        } else
+            break;
+    }
+
+    if(pPortActual)
+        *pPortActual = '\0';
+
+    /*
+    *  Rifle through the values for the host
+    */
+
+    PRAddrInfo *ai;
+    void *iter;
+    PRNetAddr addr;
+    int family = PR_AF_INET;
+
+    ai = PR_GetAddrInfoByName(hostName, PR_AF_UNSPEC, PR_AI_ADDRCONFIG);
+    if (ai) {
+        printf("%s\n", PR_GetCanonNameFromAddrInfo(ai));
+        iter = NULL;
+        while ((iter = PR_EnumerateAddrInfo(iter, ai, 0, &addr)) != NULL) {
+            family = PR_NetAddrFamily(&addr);
+            break;
+        }
+        PR_FreeAddrInfo(ai);
+    }
+
+    PSHttpServer server(host_port, family);
  
     PSHttpRequest request( &server, uri, HTTP11, 0 );
     _request = &request;
--- ./esc/src/app/xul/esc/application.ini.fix7	2009-06-22 15:34:32.000000000 -0700
+++ ./esc/src/app/xul/esc/application.ini	2009-06-22 15:35:11.000000000 -0700
@@ -25,7 +25,7 @@
 Name=ESC
 ;
 ; This field specifies your application's version.  This field is optional.
-Version=1.1.0-7
+Version=1.1.0-9
 ;
 ; This field specifies your application's build ID (timestamp).  This field is
 ; required.
